### PR TITLE
feat: legend key (DHIS2-11239)

### PR DESCRIPTION
**Pull request by martinkrulltott**
Implements [DHIS2-11239](https://jira.dhis2.org/browse/DHIS2-11239)

**Requires https://github.com/dhis2/analytics/pull/951**

**NOTE!** New lingo (as seen in the [design spec](https://docs.google.com/presentation/d/1-GrRmYy1LUNJ6I1KUL6QbOHDMbIygbym24VwMSrqs44/edit#slide=id.p))
image

---

### Key features

1. app: new option to toggle the new `legend key` from Analytics (on the right side)
2. plugin: new button to toggle the `legend key`, only visible in Dashboards
3. `legend` (bottom of chart) renamed to `seriesKey`
4. `seriesKey` shows multiple bullets for items with a legend
5. legend related props renamed to match the new backend format ([DHIS2-11347](https://jira.dhis2.org/browse/DHIS2-11347) and breaking changes in https://github.com/dhis2/analytics/pull/951)
6. legend strategy and style props imported from Analytics (`LEGEND_DISPLAY_STRATEGY_BY_DATA_ITEM`, `LEGEND_DISPLAY_STRATEGY_FIXED`, `LEGEND_DISPLAY_STYLE_FILL`, `LEGEND_DISPLAY_STYLE_TEXT`)

---

### Description

This PR revolves around legend key and series key, but grew to become quite extensive since there are a lot of moving parts. Main takeaway:
Column/bar, PT, SV and Gauge all support legends and all have the option to show a legend key.
Column/bar has a series key, which now show colors based on the legend applied.
The legend key is displayed differently in DV and on a dashboard, in DV it's an option in the Options modal, while in Dashboards there's a new button that sits at the top right corner that toggle it on and off.
To make things even more complex, the backend format for saving all of this was changed to facilitate our new terminology ([DHIS2-11347](https://jira.dhis2.org/browse/DHIS2-11347)), so there are a lot of changes to accommodate the new format.

---

### TODO

-   [x] Proper buttons for the dashboard to toggle the legend key - awaiting design by @cooper-joe 
-   [x] Cypress tests
-   [x] Hard-coded widths in VisualizationPlugin.js

---

### Screenshots

image

_new legend key (right) and multi coloured series key (bottom)_

image

_the plugin has a button to toggle the legend key that's only visible on a dashboard_

### CODE REVIEW

File: packages/plugin/src/VisualizationPlugin.js [hunk 1]
```diff
@@ -198,36 +205,72 @@ export const VisualizationPlugin = ({
...
         : null
 
+    const hasLegendSet = fetchResult.legendSets?.length > 0
+    const hasLegendKey = fetchResult.visualization.legend?.showKey
+    const transformedStyle =
+        forDashboard && hasLegendSet
+            ? {
+                  ...style,
+                  width:
+                      style.width - (showLegendKey || hasLegendKey ? 195 : 75), // TODO: Very arbitrary and static, change these values to something dynamic
```
Comment (martinkrulltott) [inline packages/plugin/src/VisualizationPlugin.js:215]: TODO: Needs to be changed

**Review by jenniferarnesen — APPROVED** [review]

**Comment by jenniferarnesen** [comment]
Approved the cypress tests

File: packages/plugin/src/VisualizationPlugin.js [hunk 2]
```diff
@@ -198,37 +205,101 @@ export const VisualizationPlugin = ({
...
+    const hasLegendSet = fetchResult.legendSets?.length > 0
+    const transformedStyle =
+        forDashboard && hasLegendSet
+            ? {
+                  ...style,
+                  width: style.width - (showLegendKey ? 200 : 36), // TODO: Very arbitrary and static, change these values to something dynamic
+              }
+            : style
+
+    const getLegendKey = () => {
```
Comment (edoardo) [inline packages/plugin/src/VisualizationPlugin.js:217]: Perhaps we can move this into a separate file (like we do with `ContextualMenu`)?

Comment (martinkrulltott) [reply packages/plugin/src/VisualizationPlugin.js:219]: Good point, this bloats VisualizationPlugin.js and should ideally be moved. I tried moving it, but it created an issue with the width calculation of the chart, which pushes the legend key out of bounds. Let's keep it as is for now and revisit this in a future refactor session.

**Pull request merged**